accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,309 @@
+
+* Cluster upgrade: As every cluster has its peculiarities when upgrading the cluster nodes and Uyuni/SUSE Manager should be agnostic about the implementation strategy, Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.upgrade_cluster <cluster name> <optional params>
+  ```
+
+### Single cluster node management with Uyuni/SUSE Manager
+
+The following actions refer to the individual cluster node registered to Uyuni/SUSE Manager. The following actions must be implemented after the cluster overview and its corresponding actions are implemented.
+NOTE: registering an SES 6 node to Uyuni/SUSE Manager is not be possible as SES 6 management node is a Salt master. From SES 7 on (under development), Salt will be used for deployment only.
+
+#### Relaxing system lock
+
+In the cluster overview, when listing individual nodes of the cluster, Uyuni/SUSE Manager will also offer the possibility of registering an individual node to Uyuni/SUSE Manager (with System > Bootstrapping).
+
+It is common between all analyzed cluster products that rebooting and package modification/removal are two forbidden actions on all nodes of the cluster.
+
+By default, Uyuni/SUSE Manager will put the minion in system lock when one of the following product is detected upon registration:
+
+* CaaSP: /etc/products.d/caasp.prod + pattern `pattern-caasp-Node`
+* SES: /etc/products.d/ses.prod
+* SLE-HA/SHAP: /etc/products.d/SLES_SAP.prod, /etc/products.d/sle-ha.prod
+
+When a system is locked, no action will be executed. System lock may be disabled by the user by disabling the corresponding Salt Formula.
+
+It will be possible to define an allowed list of actions that Uyuni/SUSE Manager can issue on a specific node by implementing a combination of selective blackout (https://github.com/uyuni-project/uyuni-rfc/pull/31/) and package locking.
+
+A product can define a list of:
+- an allowed list of actions in Uyuni/SUSE Manager action terms
+- packages or patterns to lock at the minion level
+
+This information must be provided by the product and must be inserted into Uyuni/SUSE Manager database.
+Upon bootstrapping, Uyuni/SUSE Manager will issue the required package locks and the allowed list of actions will be not part of the blackout.
+Example:
+- CaaSP define the `zypper lock` the `patterns-caasp*` pattern
+- Uyuni/SUSE Manager allows package install, modify, removal actions
+
+#### Cluster node as special citizens in the minion domain
+
+All systems add-on types will be disabled for installation, except for the Monitoring add-on.
+When deleting a system, a cleanup of salt-minion must not be done (requested by SES).
+
+All other actions will be inhibited by the system lock and will require special handling by being implemented in the cluster provider manager level.
```

Comment:
Example: node reboot: the cluster provider manager knows how to properly reboot a node:

- CaaSP: cordon and drain the node and then reboot. Or, create kured sentinel file https://github.com/SUSE/skuba/blob/master/skuba-update/skuba_update/skuba_update.py#L43
- SES: it may not be not possible
- SLE-HA: set maintenance node, reboot, reset maintenance mode.

Clarification added.